subscription.go

```diff
@@ -0,0 +1,219 @@
+	err := s.c.Send(req, func(v interface{}) error {
+		return safeAssign(v, &res)
+	})
+	if err != nil {
+		return err
+	}
+	if res.ResponseHeader.ServiceResult != ua.StatusOK {
+		return res.ResponseHeader.ServiceResult
+	}
+
+	return nil
+}
+
+func (s *Subscription) Monitor(ts ua.TimestampsToReturn, items ...*ua.MonitoredItemCreateRequest) (*ua.CreateMonitoredItemsResponse, error) {
+	// Part 4, 5.12.2.2 CreateMonitoredItems Service Parameters
+	req := &ua.CreateMonitoredItemsRequest{
+		SubscriptionID:     s.SubscriptionID,
+		TimestampsToReturn: ts,
+		ItemsToCreate:      items,
+	}
+
+	var res *ua.CreateMonitoredItemsResponse
+	err := s.c.Send(req, func(v interface{}) error {
+		return safeAssign(v, &res)
+	})
+	return res, err
+}
+
+func (s *Subscription) Unmonitor(monitoredItemIDs ...uint32) (*ua.DeleteMonitoredItemsResponse, error) {
+	req := &ua.DeleteMonitoredItemsRequest{
+		MonitoredItemIDs: monitoredItemIDs,
+		SubscriptionID:   s.SubscriptionID,
+	}
+	var res *ua.DeleteMonitoredItemsResponse
+	err := s.c.Send(req, func(v interface{}) error {
+		return safeAssign(v, &res)
+	})
+	return res, err
+}
+
+func publishOnceForSubscription(s *Subscription, acks []*ua.SubscriptionAcknowledgement) (*ua.PublishResponse, error) {
+	if acks == nil {
+		acks = []*ua.SubscriptionAcknowledgement{}
+	}
+	req := &ua.PublishRequest{
+		SubscriptionAcknowledgements: acks,
+	}
+
+	var res *ua.PublishResponse
+	err := s.c.sendWithTimeout(req, s.getPublishTimeout(), func(v interface{}) error {
+		return safeAssign(v, &res)
+	})
+	return res, err
+}
+
+func (s *Subscription) getPublishTimeout() time.Duration {
+	keepaliveIntervalMs := uint64(s.RevisedPublishingInterval) *
+		uint64(s.RevisedMaxKeepAliveCount)
+	if keepaliveIntervalMs > math.MaxUint32 {
+		keepaliveIntervalMs = math.MaxUint32
+	}
+	requestedTimeout := time.Duration(keepaliveIntervalMs) * time.Millisecond
+	if requestedTimeout < s.c.cfg.RequestTimeout {
+		return s.c.cfg.RequestTimeout
+	}
+	return requestedTimeout
+}
+
+// Run() starts an infinite loop that sends PublishRequests and delivers received
+// notifications to registered Subscriptions.
+func (s *Subscription) Run(ctx context.Context) {
+	var acks []*ua.SubscriptionAcknowledgement
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			res, err := publishOnceForSubscription(s, acks)
+			if err != nil {
+				// StatusBadSequenceNumberUnknown means at least one ack has been submitted repeatedly
+				// Ignore the error. Acks will be cleared below
+				if err == ua.StatusBadSequenceNumberUnknown ||
+					// Ignore StatusBadTimeout. No need to do anything except continue the loop
+					err == ua.StatusBadTimeout ||
+					// Ignore StatusBadNoSubscription as probably the cause is that all subscriptions
+					// are already deleted, but the publishing loop is still running and will be stopped shortly
+					err == ua.StatusBadNoSubscription {
+				} else {
+					s.c.notifySubscriptions(ctx, err)
```

Comment:
Why is this subscription notifying all other subscriptions?

Reply:
Because Publish requests are not tied to a particular subscription and you can't know, in case of an error, which subscription(s) need to know about the error. At least if you don't have a SubscriptionID in the response. Which gives me an idea that we could check for SubscriptionID and only notify all subscriptions if we really can't pinpoint one.